### ISSUE TITLE
Masker family reloaded: NiftiSpheresMasker silently ignores voxels

### ISSUE THREAD

**Issue by banilo** [nilearn/nilearn#1638]
Using the NiftiSpheresMasker() I found out: If no mask_img is specified and it is thus computed by  a heuristic on the data images fed into transform(), then spheres that touch *any* voxel not inside the automatically computed mask will be set to overall NaN. That is, even if most of the sphere is inside the data space, the overall sphere feature appears to result to NaN if a tiny part is not inside the implicit mask. In my particular case, it took me some time to find this source of unexpected behavior, and it was fixed my explicitly specifying a liberal mask as "search space".

Perhaps it would be good to add an option that controls this behavior more directly and add a warning to the output that informs the user why some values in the sphere space will turn out to be NaN coming out of transform().

**Comment by AlexandreAbraham** [comment]
If I understand well your usecase: you use NiftiSpheresMasker. When you do not put a mask_img, you get NaN for some spheres. When you put a mask, you get values in these spheres.

First, there is no heuristic to compute masks automatically in the NiftiSpheresMasker. So my guess here is that, in your data, you must have NaNs outside of the brain and, if a sphere overlaps without outside brain voxels, the mean ends up being a NaN. When you put the mask, NaNs are removed, and then you get non-NaN values.

Could you check is you have NaNs outside of the brain? This would allow us to rule out this case.

Sidenote: we should probably detect voxels with NaNs and raise an error if they are not masked.

**Comment by himanshupathak21061998** [comment]
I want to work on this Issue . According to this if the value mask_img argument is None than I have to give a warning message and detect those voxels which are turned into NaN . This much is sufficient or anything else is required

**Comment by himanshupathak21061998** [comment]
I want to work on this Issue . In this issue if the mask_img parameter is None in class NiftiSpheresMasker then I have to give a warning message and detect those voxels which are turned into Nan . This much is sufficient or anything else I have to do .

**Comment by himanshupathak21061998** [comment]
Please I need some suggestions help me @KamalakerDadi

**Comment by KamalakerDadi** [comment]
Hi @himanshupathak21061998 Thanks for your interest. Let me get back to this with a reproducible script and then we fix it.
Thanks again

**Comment by KamalakerDadi** [comment]
@banilo Can you provide Nifti files to reproduce ? Thanks a lot!

**Comment by kimberlylray** [comment]
Hi - following on a similar issue I recently ran into with NiftiSpheresMasker. 
Im extracting the time series from spherical nodes in the Power atlas.  Some Power nodes are on the edge of the brain and thus may extend outside of the brain depending on their size. 
Is it possible to extract the % of voxels in a sphere that contain signal? Mainly, I want to ensure that a sphere on the edge of the brain contains enough voxels with signal to provide a reliable estimate. 
Many thanks! 
@dpys

**Comment by GaelVaroquaux** [comment]
> Is it possible to extract the % of voxels in a sphere that contain signal?

Apply extraction to a 0/1 mask? That should give you the number that you
are interested in.

**Comment by KamalakerDadi** [comment]
@kimberlylray Could you provide us a fmri image uploading them some where ?

This would help us to take a look more closely. Also, if you could tell us the radius you have specified.

Thanks a lot!

**Comment by dPys** [comment]
Hi @kimberlylray 

Based on @GaelVaroquaux 's answer, maybe something like a dictionary would be useful?

```
import numpy as np
from nilearn.masking import compute_epi_mask
from nilearn import input_data

mask_img = compute_epi_mask(func_file)
ts_within_nodes = spheres_masker.fit_transform(func_file)
mask_within_nodes = spheres_masker.transform(mask_img)
seeds = [(round(s[0], 2), round(s[1], 2), round(s[2], 2)) for s in spheres_masker.seeds]
coverage_dict = dict(zip(seeds, mask_within_nodes[0]))

if any(np.isnan(val) for val in coverage_dict.values()):
    print("Warning: null values detected in spheres")
elif any(float(val) < 1.0 for val in coverage_dict.values()):
    print("Warning: at least one sphere contains missing voxels")
else:
    pass
```

I can say that in PyNets, one thing I've recently implemented before running NiftiSpheresMasker is a function to prune away coordinates for spheres that are not included in the brain (even for whole-brain atlases, this can actually occur quite often). Referencing https://github.com/dPys/PyNets/blob/fe8e55d453e9d73623e3a84acad146d3dbb37d9e/pynets/nodemaker.py#L292 might offer some ideas.

My 2 cents,
@dPys

**Comment by kimberlylray** [comment]
Thank you @GaelVaroquaux for the suggestion, you solution seems quite obvious now.

Thanks for your response @KamalakerDadi, I attached an ROI with radius 6mm in MNI space that barely extends outside the subject brain mask. The brain mask is an output from fMRIprep.
[sub-13138_space-MNI152NLin2009cAsym_desc-brain_mask.nii.gz](https://github.com/nilearn/nilearn/files/2634020/sub-13138_space-MNI152NLin2009cAsym_desc-brain_mask.nii.gz)

[Schaefer2018_400Parcels_7Networks_order_FSLMNI152_2mm_roi182.nii.gz](https://github.com/nilearn/nilearn/files/2634007/Schaefer2018_400Parcels_7Networks_order_FSLMNI152_2mm_roi182.nii.gz)

**Comment by himanshupathak21061998** [comment]
@KamalakerDadi Can we start working on this issue . I think you have got nifti files provided by @kimberlylray So we can start working . Please give me some pointer on how should I start my approach to solve this issue Just I need some elaboration to start  working

**Comment by KamalakerDadi** [comment]
>@KamalakerDadi Can we start working on this issue .

Yes. That's planned in coming week.

**Comment by KamalakerDadi** [comment]
I took the seeds as given by @kimberlylray as they lies on the edge of the brain. As you can see here:
![seed_on_mask_image](https://user-images.githubusercontent.com/11410385/50443245-a5566d00-0902-11e9-916e-2ace5afc841a.png)

We don't have problem of reporting nans when outer layer of the brain has no nans. This example script shows that, even after increasing the size of the radius.
```python
from nilearn import datasets
from nilearn.input_data import NiftiSpheresMasker
from nilearn.plotting import find_parcellation_cut_coords

adhd = datasets.fetch_adhd(n_subjects=1)
seeds = find_parcellation_cut_coords('Schaefer2018_400Parcels_7Networks_order_FSLMNI152_2mm_roi182.nii.gz')
# radius of size 3mm without mask image
spheres_masker = NiftiSpheresMasker(seeds=seeds, radius=3.)
signals1 = spheres_masker.fit_transform(adhd.func[0])

# radius of size 50 without mask image
spheres_masker = NiftiSpheresMasker(seeds=seeds, radius=50.)
signals2 = spheres_masker.fit_transform(adhd.func[0])
```

Problem comes when the given brain images has ```np.nan``` outside of the brain or may be inside the brain. As this example script outputs nans when spheres are interacted with the nans which are lying outside of the brain. This interaction would happen when there is an increase in spheres size (radius).
```python
import numpy as np
import nibabel

from nilearn.input_data import NiftiSpheresMasker

affine = np.eye(4)

data_with_nans = np.random.random((10, 10, 10))
data_with_nans[:, :, :] = np.nan

data_without_nans = np.random.random((9, 9, 9))
indices = np.nonzero(data_without_nans)

# Leaving nans outside of some data
data_with_nans[indices] = data_without_nans[indices]
img = nibabel.Nifti1Image(data_with_nans, affine)

mask = np.ones((9, 9, 9))
mask_img = nibabel.Nifti1Image(mask, affine)

seed = [(7, 7, 7)]
# No interaction of seed with nans
masker = NiftiSpheresMasker(seeds=seed, radius=1.)
print(masker.fit_transform(img))

# Interaction of seed with nans
masker = NiftiSpheresMasker(seeds=seed, radius=2.)
print(masker.fit_transform(img))

# When mask image is provided, the seed interacts within the brain, so no nans.
masker = NiftiSpheresMasker(seeds=seed, radius=2., mask_img=mask_img)
print(masker.fit_transform(img))
```
This helps in to see where in the code we can raise ```ValueError``` or ```warnings.warn``` to the end-user.
Will get back to you @himanshupathak21061998 to fix this.

**Comment by bthirion** [comment]
I'd rather go for the first one. SPM, for instance, introduces Nans for non-existant data, which is an issue.
Padding with zeros is not great, but it's acceptable I think, and it is less disturbing than raising errors.

**Comment by KamalakerDadi** [comment]
@bthirion agreed on the first one. Can you do a PR ? @himanshupathak21061998

**Comment by KamalakerDadi** [comment]
Closing with #1906
